### ISSUE TITLE
Struggling with passwordless SSH

### ISSUE THREAD

**Issue by unfazing** [horovod/horovod#2669]
**Environment:**
1. Framework: (TensorFlow, Keras, PyTorch, MXNet)
2. Framework version: 
3. Horovod version:
4. MPI version:
5. CUDA version:
6. NCCL version:
7. Python version:
8. OS and version:
9. GCC version:

I am using a docker image, horovod/horovod:0.20.0-tf2.3.0-torch1.6.0-mxnet1.6.0.post0-py3.7-cuda10.1, pulled from docker hub on machines running Ubuntu 20.04

**Checklist:**
1. Did you search issues to find if somebody asked this question before? yes
2. If your question is about hang, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/running.rst)? 
3. If your question is about docker, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/docker.rst)? yes
4. Did you check if you question is answered in the [troubleshooting guide](https://github.com/horovod/horovod/blob/master/docs/troubleshooting.rst)? yes

**Your question:**
I am able to SSH into the secondary worker's host through port 22 from my primary worker's container. But I am unable to SSH into the secondary worker's container through port 12345; it requests a password and does not accept the offered public key.

I didnt quite understand the /mnt/share portion of the tutorial in https://horovod.readthedocs.io/en/stable/docker.html

So I instead mounted the local path -v /home/user/.ssh:/root/.ssh when running the docker container for both the primary and secondary workers. I carried out the steps linked in here http://www.linuxproblem.org/art_9.html to allow for passwordless ssh.

My question is if there is another way to go about doing this without the shared filesystem; and what should be present in the directory that I mount to /root/.ssh if it is a share filesystem.

**Comment by tgaddair** [comment]
Hey @unfazing, the shared filesystem just makes the example simpler, but the important part is that the SSH keys are available in all the worker containers.  This can be done by baking them into the container, rsyncing them across workers, etc.

The content of this folder is described here: http://www.linuxproblem.org/art_9.html

**Comment by unfazing** [comment]
Hello, after trying to resync the id_rsa and authorized_keys file from within the workers, I now face a different error message when running this code on the pri worker's container (x.x.x.x.5 is the sec worker's ip):

`ssh x.x.x.x.5 -p 12345 -vvvv`

error message:
> debug1: Connecting to x.x.x.x.5 [x.x.x.x.5] port 12345.
> debug1: Connection established.
> debug1: permanently_set_uid: 0/0
> debug1: identity file /root/.ssh/id_rsa type 0
> debug1: key_load_public: No such file or directory
> debug1: identity file /root/.ssh/id_rsa-cert type -1
> debug1: key_load_public: No such file or directory
> debug1: identity file /root/.ssh/id_dsa type -1
> debug1: key_load_public: No such file or directory
> debug1: identity file /root/.ssh/id_dsa-cert type -1
> debug1: key_load_public: No such file or directory
> debug1: identity file /root/.ssh/id_ecdsa type -1
> debug1: key_load_public: No such file or directory
> debug1: identity file /root/.ssh/id_ecdsa-cert type -1
> debug1: key_load_public: No such file or directory
> debug1: identity file /root/.ssh/id_ed25519 type -1
> debug1: key_load_public: No such file or directory
> debug1: identity file /root/.ssh/id_ed25519-cert type -1
> debug1: Local version string SSH-2.0-OpenSSH_7.6p1 Ubuntu-4ubuntu0.3
> ssh_exchange_identification: Connection closed by remote host

**Comment by tgaddair** [comment]
Hey @unfazing, I would suggest reading up on some SSH specific threads like this one to debug this: https://unix.stackexchange.com/questions/128894/ssh-exchange-identification-connection-closed-by-remote-host-not-using-hosts-d

**Comment by unfazing** [comment]
Thank you for your help. Following your link, I used the `$(which sshd) -Dddp 12345` command on the worker container to have the sshd listen to the alternative port 12345 and display its verbose debug output in the terminal. From there I saw the following error message when attempting to authenticate with the public key:

> Authentication refused: bad ownership or modes for file /root/.ssh/authorized_keys

Referencing [this thread](https://github.com/panubo/docker-sshd/issues/32), I found that my mistake was binding the host ssh keys to the container's /root/.ssh path. Rather I should have copied the keys into the container directly. It works now. Thank you.
